deducing-this.md

```diff
@@ -295,17 +392,20 @@ class string_builder {
     }
     return self;
   }
+
   string_builder bop(string_builder this self) {
     s.append("bop");
     return self;
   }
 };
 
 // this is optimally efficient as far as allocations go
-std::string const x = (string_builder{{"asdf"}} * 5).bop().bop();
+std::string const x = (string_builder{"asdf"} * 5).bop().bop();
 ```
 
-Of course, implementing this example with templated `this` member functions would have been slightly more efficient due to also saving on move constructions, but the by-value `this` usage makes for simpler code.
+In this example, `x` would hold the value `"asdfasdfasdfasdfasdfbopbop"`.
+
+Of course, implementing this example with templated explicit object parameter member functions would have been slightly more efficient due to also saving on move constructions, but the by-value usage makes for simpler code.
```

Comment:
The by-value builder method is also the only way you can pass a function pointer and still be able to use it by either move or copy constructing the parameter once you type-erase it - you can't do that with templates.

Can't really figure out how to word that properly but it's basically "if you want to type-erase a builder interface, the only way to do it is with pass-by-value methods, since templates are not virtualizable.

Reply:
I'm not sure I understand this. Can you give me an example?

Reply:
Pushed corrections for all the comments but this one.

Reply:
Lemme see:

```cpp
struct builder {
    std::string buffer;
    builder beep(builder this self) { return {buffer + "beep"}; }
    builder boop(builder this self) { return {buffer + "boop"}; }
};
using m_type = decltype(builder::beep);
static_assert(std::is_same(m_type, decltype(builder::boop)));

int boop_the_builder(m_type builder_xform) {
    // imagine this is shared_ptr<builder> and do_i_own is "ref-count is 1"
    auto&& [do_i_own, this_builder] = get_builder();
    if (do_i_own) {
        return builder_xform(std::move(this_builder));
    } else {
        return builder_xform(this_builder); // copy
    }
}

int main() {
    std::cout << boop_the_builder(builder::beep);
    std::cout << boop_the_builder(builder::beep);
}
```